server.go

```diff
@@ -543,102 +542,147 @@ func (s *server) establishPersistentConnections() error {
-		return err
-	case <-s.quit:
-		return errors.New("server shutting down")
+			srvrLog.Debugf("Skipping %v in broadcast",
+				skip.SerializeCompressed())
+
+			continue
+		}
+
+		// Dispatch a go routine to enqueue all messages to this peer.
+		s.wg.Add(1)
+		wg.Add(1)
+		go s.sendPeerMessages(sPeer, msgs, &wg)
 	}
-}
 
-// sendReq is  message sent to the server by a related subsystem which it
-// wishes to send a set of messages to a specified peer.
-type sendReq struct {
-	target *btcec.PublicKey
-	msgs   []lnwire.Message
+	// Wait for all messages in this broadcast to be enqueued.
+	wg.Wait()
 
-	errChan chan error
+	return nil
 }
 
 type nodeAddresses struct {
 	pubKey    *btcec.PublicKey
 	addresses []*net.TCPAddr
 }
 
-// sendToPeer send a message to the server telling it to send the specific set
+// SendToPeer send a message to the server telling it to send the specific set
 // of message to a particular peer. If the peer connect be found, then this
 // method will return a non-nil error.
-func (s *server) sendToPeer(target *btcec.PublicKey, msgs ...lnwire.Message) error {
-	errChan := make(chan error, 1)
-
+// NOTE: This function is safe for concurrent access.
+func (s *server) SendToPeer(target *btcec.PublicKey, msgs ...lnwire.Message) error {
 	msgsToSend := make([]lnwire.Message, 0, len(msgs))
 	msgsToSend = append(msgsToSend, msgs...)
-	sMsg := &sendReq{
-		target:  target,
-		msgs:    msgsToSend,
-		errChan: errChan,
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	return s.sendToPeer(target, msgsToSend)
+}
+
+// sendToPeer is an internal method that delivers messages to the specified
+// `target` peer.
+func (s *server) sendToPeer(target *btcec.PublicKey, msgs []lnwire.Message) error {
+	// Compute the target peer's identifier.
+	targetPubBytes := target.SerializeCompressed()
+
+	srvrLog.Infof("Attempting to send msgs %v to: %x",
+		len(msgs), targetPubBytes)
+
+	// Lookup intended target in peersByPub, returning an error to the caller if
+	// the peer is unknown.  Access to peersByPub is synchronized here to ensure
+	// we consider the exact set of peers present at the time of invocation.
+	targetPeer, ok := s.peersByPub[string(targetPubBytes)]
+	if !ok {
+		srvrLog.Errorf("unable to send message to %x, "+
+			"peer not found", targetPubBytes)
+
+		return errors.New("peer not found")
 	}
 
-	select {
-	case s.sendRequests <- sMsg:
-	case <-s.quit:
-		return errors.New("server shutting down")
+	s.sendPeerMessages(targetPeer, msgs, nil)
+
+	return nil
+}
+
+// sendPeerMessages enqueues a list of messages into the outgoingQueue of the
+// `targetPeer`.  It accepts an additional `wg` parameter so that a single
+// broadcast can be synchronized.
+// NOTE: This method MUST provide a non-nil `wg` if it is spawned as a go
+// routine, otherwise `wg` should be nil.
+func (s *server) sendPeerMessages(
+	targetPeer *peer,
+	msgs []lnwire.Message,
+	wg *sync.WaitGroup) {
+
+	if wg != nil {
+		defer wg.Done()
+		defer s.wg.Done()
```

Comment:
I was mistaken, there were some flakes that were caused by not having synchronous broadcast (was the original intent of the PR after all 😂). I've updated the docs to hopefully alleviate the ambiguity